sdk/samples/iot/hub/src/paho_iot_hub_pnp_example.c

```diff
@@ -0,0 +1,744 @@
+    az_span_copy(buffer, env_span);
+    *out_value = az_span_slice(buffer, 0, az_span_size(env_span));
+  }
+  else if (default_value != NULL)
+  {
+    printf("%s\n", default_value);
+    az_span default_span = az_span_from_str(default_value);
+    AZ_RETURN_IF_NOT_ENOUGH_SIZE(buffer, az_span_size(default_span));
+    az_span_copy(buffer, default_span);
+    *out_value = az_span_slice(buffer, 0, az_span_size(default_span));
+  }
+  else
+  {
+    printf("(missing) Please set the %s environment variable.\n", env_name);
+    return AZ_ERROR_ARG;
+  }
+
+  return AZ_OK;
+}
+
+// Create mqtt endpoint e.g: ssl//contoso.azure-devices.net:8883
+static az_result create_mqtt_endpoint(char* destination, int32_t destination_size, az_span iot_hub)
+{
+  int32_t iot_hub_length = (int32_t)strlen(iot_hub_hostname);
+  int32_t required_size = az_span_size(mqtt_url_prefix) + iot_hub_length
+      + az_span_size(mqtt_url_suffix) + (int32_t)sizeof(null_terminator);
+
+  if (required_size > destination_size)
+  {
+    return AZ_ERROR_INSUFFICIENT_SPAN_SIZE;
+  }
+
+  az_span destination_span = az_span_init((uint8_t*)destination, destination_size);
+  az_span remainder = az_span_copy(destination_span, mqtt_url_prefix);
+  remainder = az_span_copy(remainder, az_span_slice(iot_hub, 0, iot_hub_length));
+  remainder = az_span_copy(remainder, mqtt_url_suffix);
+  az_span_copy_u8(remainder, null_terminator);
+
+  return AZ_OK;
+}
+
+// Read the user environment variables used to connect to IoT Hub
+static az_result read_configuration_and_init_client()
+{
+  az_span cert = AZ_SPAN_FROM_BUFFER(x509_cert_pem_file);
+  AZ_RETURN_IF_FAILED(
+      read_configuration_entry(ENV_DEVICE_X509_CERT_PEM_FILE, NULL, false, cert, &cert));
+
+  az_span trusted = AZ_SPAN_FROM_BUFFER(x509_trust_pem_file);
+  AZ_RETURN_IF_FAILED(
+      read_configuration_entry(ENV_DEVICE_X509_TRUST_PEM_FILE, "", false, trusted, &trusted));
+
+  az_span device_id_span = AZ_SPAN_FROM_BUFFER(device_id);
+  AZ_RETURN_IF_FAILED(
+      read_configuration_entry(ENV_DEVICE_ID, "", false, device_id_span, &device_id_span));
+
+  az_span iot_hub_hostname_span = AZ_SPAN_FROM_BUFFER(iot_hub_hostname);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      ENV_IOT_HUB_HOSTNAME, "", false, iot_hub_hostname_span, &iot_hub_hostname_span));
+
+  // Paho requires that the MQTT endpoint be of the form ssl://<HUB ENDPOINT>:8883
+  AZ_RETURN_IF_FAILED(
+      create_mqtt_endpoint(mqtt_endpoint, (int32_t)sizeof(mqtt_endpoint), iot_hub_hostname_span));
+
+  // Initialize the hub client with the hub host endpoint and the default connection options
+  AZ_RETURN_IF_FAILED(az_iot_hub_client_init(
+      &client,
+      az_span_slice(iot_hub_hostname_span, 0, (int32_t)strlen(iot_hub_hostname)),
+      az_span_slice(device_id_span, 0, (int32_t)strlen(device_id)),
+      NULL));
+
+  return AZ_OK;
+}
+
+static az_result report_method(az_span payload, az_span response, az_span* out_response)
```

Comment:
I don't think `report_method` captures what this is doing since report to me implies I/O from device=>Cloud, especially for someone coming into this .c file without carefully studying the DTDL to understand its implementing the report.

I think this is really like `invoke_getMaxMinReport`?

Reply:
agreed changing